Use Settlement Contract's Internal Buffer to Optimize Unwraps

MartinquaXD opened the pull request:
Fixes #1496

This change will cause the internal buffers of the settlement contract to be used for unwraps if somebody buys ETH in a batch.
Before this change, every ETH buy order would result in an unwrap operation during the settlement. With this change we try to buffer unwraps such that we unwrap less frequently. But when we do, we unwrap way more than the current settlement needs so we don't have to unwrap again for a few settlements.
This basically uses the same idea as a `Vec<T>` trying to reduce the number of reallocations when you push elements into it.

If the settlement's unwraps can be paid with the settlement contract's ETH buffer, it does so and simply drops the unwrap operation of the settlement. If that is not possible, the driver computes a WETH amount to unwrap given a user defined factor. If that unwrap would exceed what the current settlement needs, the driver executes the unwrap. If that would NOT yield enough ETH for the settlement, the driver does nothing and leaves the settlement as is which results in the behavior we had before this PR.

The user can define how much of the WETH buffer may be unwrapped by passing `--weth-unwrap-factor` to the solver. This value is supposed to be in the range of [0.0, 1.0]. The unwrap amount is equal to `ETH_BALANCE * weth_unwrap_factor`. I followed the recommendation of 0.6 as the default value like it has been suggested in the linked issue.

For more information about the issue, have a look at the linked issue. It is quite detailed.

### Test Plan
Added unit tests for the algorithm optimizing unwraps.

Did a manual test trading GNO -> ETH on rinkeby without unwraps: https://rinkeby.etherscan.io/tx/0xf326ce6ce3d2da376d233f73d4d9646054eff0d91181825ba6eb8512f98003b5

Commit:
Add CLI option to configure WETH unwrap factor

--weth-unwrap-factor determines how much of the settlement contract's
WETH reserves may be unwrapped at once in order to reduce the cost of
unwraps per settlement.

Commit:
Reduce cost of unwrap per settlement

So far every settlement involving a buy order for ETH involved an unwrap
operation. To reduce the number of unwraps we can use the settlement
contract's ETH buffer to pay out the ETH and if that runs low, a big
portion of the settlement contract's WETH buffer can be unwrapped at once.
Because we unwrap WETH less frequently the cost of unwraps per
settlement measured over multiple settlements also decreases.

According to Ben this could cut our cost by up to 3% in total.

Because we can't really determine how much of the WETH buffer the solver
planned to use in its suggested solution, we have to simulate the
settlement after an optimization has been applied. Otherwise the
optimization could cause a settlement to fail.

To avoid simulating unneccessarily I added early returns whenever
they made sense.

MartinquaXD commented:
One thing I'm unhappy about is the introduction of the `shutdown_sender`. If anybody has a good idea how to ditch that, let me know.

bh2smith commented on crates/e2e/tests/e2e/onchain_settlement.rs:
```diff
@@ -136,6 +136,7 @@ async fn onchain_settlement(web3: Web3) {
         block_stream,
         solvable_orders_cache,
         base_tokens,
+        ..
```
I guess this is the rust notation for a partial destructure.

bh2smith commented on crates/e2e/tests/e2e/unwrap_cost_optimization.rs:
```diff
@@ -0,0 +1,388 @@
+use crate::deploy::Contracts;
```
Looks like the content of this file was moved from elsewhere. If there is any new code here it's difficult to tell....

MartinquaXD replied:
I started with the eth_integration tests as the basis for this new test but had to rewrite quite a bit.

bh2smith reviewed (commented):
Looks pretty good but that one new file is difficult to read... is that all new code?

sunce86 commented on crates/e2e/tests/e2e/unwrap_cost_optimization.rs:
```diff
@@ -0,0 +1,388 @@
+    (solver, trader_a, trader_b, token)
+}
+
+async fn set_up_driver_and_services(
```
Could we avoid calling `OrderbookServices::new` twice for one test by separating this function to two functions? Is it necessary to have two `OrderbookServices` objects in one test?

sunce86 reviewed (approved):
Looks good. Well explained PR.

fleupold commented on crates/shared/src/arguments.rs:
```diff
@@ -147,6 +147,12 @@ pub fn parse_fee_factor(s: &str) -> Result<f64> {
     Ok(fee)
 }
 
+pub fn parse_unwrap_factor(s: &str) -> Result<f64> {
```
nit: My initial thought was "why don't we reuse the same parsing as for the fee factor?" (answer, because this one here is upper bound by 1).

Maybe we can rename these methods to be more explanatory on the kind of parsing they do (e.g. `parse_percentage_factor` vs `parse_unbounded_percentage_factor`) rather than which field they are used for (I would like them to be reused if we add another field)?

MartinquaXD replied:
Good suggestions, will do.

fleupold commented on crates/solver/src/driver.rs:
```diff
@@ -145,13 +148,70 @@ impl Driver {
         .await
     }
 
+    async fn optimize_cost_of_unwrapping(&self, settlement: &mut Settlement) -> Result<()> {
```
I would like to avoid adding this logic to `driver.rs` as it is already little bit our "messy cellar" where we have a lot of different logic together.
I was originally thinking that this kind of "settlement post processing" could be it's own component/pipeline where we can add multiple optimisations in a row. With this being our first one, it might be a bit too early to build an abstraction (we don't really know yet what other processing steps we will have, if any), but maybe we can move this logic into it's own submodule (settlement_post_processing or the like).

Also, I feel this method doesn't necessarily require a e2e test, but could well be covered by unit tests (given that the settlement is supposed to be pretty much a data struct and we are "just" doing some logically involved modifications to it). Currently the difficult part for a unit test is that we need to know the buffer which requires an async call on an unmockable web3 instance.

We already have a `BufferRetrieving` trait which is used by the `http_solver` to pass internal buffers into our MiP and Quasimodo solver. This trait could be extended to also support the native token (right now it's limited to ERC20) and then used here. We could then easily mock it and have unit tests covering all of the scenarios.

This is also brings the small performance advantage that we could create a cache of buffers "per settlement run" to avoid double fetching these balances (not needed for this PR).

fleupold commented on crates/solver/src/driver.rs:
```diff
@@ -145,13 +148,70 @@ impl Driver {
+        }
+
+        let weth = contracts::WETH9::at(&self.web3, self.native_token);
+        let weth_balance = weth
```
One complexity that I didn't think of before is, what if the solution is already planning to use part of the WETH buffer (we now allow MiP and Quasimodo to use internal buffers during their solution finding)? Then converting 60% of it to ETH may cause the settlement to fail. I'm not sure if we can somehow know the state of the buffers after the settlement was applied. 

It may require a simulation + balance query which is not trivial to implement without using [`trace_callMany`](https://openethereum.github.io/JSONRPC-trace-module#trace_callmany) (a feature of OpenEthereum, the node type we want to stop depending on) or building a [`StorageReaderExtension`](https://www.youtube.com/watch?v=4BfAEIp2_No) for our smart contract.

Alternatively, we could simulate the "optimised settlement" and discard if it fails or only perform this optimisation for our non-http-solvers (basically requiring solvers to either do all or none of the "internal buffer" optimisation)

MartinquaXD replied:
I assumed `SettlementEncoder::unwraps` contained all unwraps which will happen during the execution of the settlement but judging by this comment this probably only accounts for the ETH which gets paid out to the traders in the end.
Could we maybe get the needed information by analyzing `SettlementEndoder::execution_plan`? Like sum up all `Interactions` which take from our internal buffer.

fleupold replied:
Afaik, `SettlementEncoder::unwraps` does contains all unrwaps. The problem is that we don't necessarily execute all swaps the are needed for the settlement to "perfectly clear" (sample settlement: https://etherscan.io/tx/0xec7fcc9c59e3b56d29ea7feb3f79afc2fb1c0c73e7c0049075be1666a64f2ced, see how we do a bunch of trades without any external swap - the imbalance is paid from "internal buffers")

Here is an example where this new logic would fail: Let's say we have 10 WETH (0 ETH) in the contract and there is a trade selling 20k USDC for 5 WETH together with a trade selling some shitcoin (which is not "whitelisted" for internal buffer trading) for 1 ETH. 
Since recently, our http solvers receive information about the buffers as part of the auction instance and may - in order to be more gas efficient - decide to use less interactions (e.g. skip the $20k -> 5 WETH swap) and pay those from internal buffers.

In this case the logic here would decide to unwrap 6 WETH to ETH, however this wouldn't leave enough WETH to pay for the $20k -> 5 WETH trade.

Judging the expected in/out amounts from interactions would require adding the in/out token + amounts to the `Interaction` trait - something we have spoken about doing in the past (it might also be relevant for a smarter solutions submission strategy that @sunce86 is working on). I'm not sure how big of a refactoring this would be and if it would be simpler for now to simulate the optimised solution (maybe only necessary if we add an unwrap) and discard it if it's no longer valid.

fleupold commented on crates/solver/src/driver.rs:
```diff
@@ -145,13 +148,70 @@ impl Driver {
+                    amount: amount_to_unwrap,
+                });
+        } else {
+            tracing::warn!("consider topping up settlement contract's internal buffer");
```
Not needed for now, but maybe a metric by how much we would have had to top up the contract could be useful to quantify the correct amount to use on average (can maybe be done using Dune as well)

fleupold reviewed (commented):
I would really like to explore if we can get around without an e2e test. They are quite heavy weight, so we try to only test major features (new order types, settlement flows). Post processing could be considered such a major feature, but if we can test the logic in a unit test, I'd prefer that.

Also one comment regarding our HTTP solvers using buffers already partially now.

MartinquaXD commented:
Thanks for the great feedback!
It seems that the major concerns can be addressed by replacing the e2e test with a unit test. This means I don't have to manage the lifetime of the orderbook service which caused me to introduce the shutdown mechanism in the first place.
I initially wanted to have an e2e test to see whether I overlooked any relevant nuances in the interaction between the contract and the driver. Since that doesn't seem to be the case, a unit test should suffice.

MartinquaXD commented on crates/e2e/tests/e2e/services.rs:
```diff
@@ -187,12 +188,20 @@ impl OrderbookServices {
+                if shutdown_receiver.await.is_err() {
+                    // Tests dropping the shutdown_sender would terminate the service by accident.
+                    // Let's wait for an explicit shutdown message before terminating the service.
+                    std::future::pending::<()>().await;
```
The pending future we had before never got dropped. The problem started happening after I started using the oneshot channel for the shutdown mechanism.
Just replacing the `pending()` future with a future waiting for a message from a oneshot channel caused the future to terminate once the channel sender has been dropped. This meant all the existing e2e tests would have to keep the `shutdown_sender` alive. To not make every test manage the `shutdown_sender` when they never had the intention to terminate the service, I required an actual shutdown message to be sent (which can be differentiated from termination caused by dropping the sender). If `shutdown_receiver.await` terminated because the sender got dropped, it was one of the tests which didn't want to terminate the service anyway so I used `pending()` as the expected fallback behavior.
But that doesn't matter anymore anyway because most of the problems can be solved by replacing the e2e with an unit test altogether.

fleupold replied:
I see, my bad was that I didn't realise that if shutdown_receiver.await.is_ok() we also resolve the future.

Commit:
Implement unit tests

Tests 4 scenarios
1) dropping the unwrap succeeds
2) convert way more WETH than the settlment needs to pay out
3) optimizations don't apply because the buffer is too small
4) optimizations don't apply because an error occurred

MartinquaXD commented:
* created a new module for post processing `Settlements` (instead of putting it in `driver.rs` directly)
* replaced expensive e2e tests with cheap unit tests

Making the API better testable made it a little weird. Maybe I should pass `impl Trait` instead of closures in the optimization algorithm?

vkgnosis commented on crates/solver/src/settlement/settlement_encoder.rs:
```diff
@@ -340,6 +340,24 @@ impl SettlementEncoder {
+
+    /// Drops all UnwrapWethInteractions for the given token address.
+    /// This can be used in case the settlement contracts ETH buffer is big enough.
+    pub fn drop_unwrap(&mut self, token: H160) {
```
I think there can only be one "weth" address per network. So it doesn't make sense to have the token address as a parameter here. Maybe someone else could confirm that.

MartinquaXD replied:
In that case wouldn't it also make sense to make `SettlementEncoder::unwraps` of type `U256` instead of `Vec<UnwrapWethInteraction`? Or maybe `UnwrapWethInteraction` if we still need the included `WETH9` struct?

vkgnosis replied:
I think atm we need the weth9 instance so that we can implement the Interaction trait. With some refactoring we might be able to change that too and it probably would be slightly cleaner.

nlordell replied:
> I think there can only be one "weth" address per network. So it doesn't make sense to have the token address as a parameter here.

The reason for this is that there are multiple "unwrap" tokens - cDAI <-> DAI, stETH <-> ETH to name a few that we potentially wanted to support here.

That being said, I would be OK with ripping this out of the the settlement encoder in favour of a single `Unwrap` interaction per settlement since we don't use this feature and it is causing confusion ATM. That being said, what I just mentioned is unrelated to this change though.

nlordell replied:
> maybe open another PR refactoring the unwraps properly

Of course! This is what I meant with:
> what I just mentioned is unrelated to this change though

vkgnosis commented on crates/solver/src/settlement_post_processing/optimize_unwrapping.rs:
```diff
@@ -0,0 +1,201 @@
+use contracts::WETH9;
+use primitive_types::U256;
+
+pub async fn optimize_unwrapping<V, VFut, B, BFut>(
+    settlement: &mut Settlement,
```
Since we have to clone the settlement anyway maybe it would be nicer to take and return it by value. That way the caller can't ignore the result so easily and it is clearer under what conditions the settlement changes. With the current signature it might not be clear whether the settlement will be modified if the function returns Err.

MartinquaXD replied:
I used an out parameter to leave the original `Settlement` untouched while also reporting an error. With `-> Result<Settlement>` we would loose the original settlement in case of an error and with `-> Settlement` we would not be able to report any errors at all.
Because there is currently no good way to handle an error anyway, I could simply drop the error handling and always return a settlement.

vkgnosis replied:
We would have to clone it to not lose it, yes (or return it in the Error). I do think this way is fine too so no need to change it. Maybe add a comment saying that settlement is not modified if the Result is Err.

MartinquaXD replied:
I decided to drop the error propagation after all.

vkgnosis commented on crates/solver/src/settlement_post_processing/mod.rs:
```diff
@@ -0,0 +1,74 @@
+        )
+        .await;
+
+        debug_assert!(settlement_would_succeed(settlement.clone()).await);
```
I don't think this is useful because optimize_unwrapping already does it and it feels weird  to do IO in a debug assert but not sure.

MartinquaXD replied:
I thought it nicely conveys the post conditions of the `PostProcessingPipeline` in case some more stages get added to it in the future. But I would also be fine with dropping it.

nlordell replied:
My worry is that network changes between the initial simulation and the final simulation for the `debug_assert` may cause settlements to start failing.

Since this is a `debug_assert` and doesn't actually run in our deployments (we use release builds AFAIK) its not that big of an issue, but it may cause unexpected panics when running locally in debug mode.

MartinquaXD replied:
Dropped the assert.

vkgnosis commented:
> Making the API better testable made it a little weird. Maybe I should pass `impl Trait` instead of closures in the optimization algorithm?

It might be easier to extract the IO you need into a trait that can be mocked. You can look at other examples in the code where use mockall::automock.

Commit:
Rename parse_unbounded_percentage_factor -> parse_unbounded_factor

Commit:
Write comments explaining behavior if optimization returns Err

Commit:
drop debug_assert

Commit:
Drop error propagation from optimzations

Propagating errors during optimizing doesn't give us much and makes the
function signature confusing.

Commit:
Replace closure argument with impl BufferRetrieving

nlordell commented on crates/solver/src/settlement_post_processing/optimize_unwrapping.rs:
```diff
@@ -0,0 +1,206 @@
+/// If this function returns Err(), the original settlement will not be modified in any way.
+pub async fn optimize_unwrapping<V, VFut, B, BFut>(
+    settlement: &mut Settlement,
+    settlement_would_succeed: V,
+    get_weth_balance: B,
```
Cool use of closures to make the method testable instead of using a trait and mocks.

nlordell replied:
> It might be easier to extract the IO you need into a trait that can be mocked. You can look at other examples in the code where use mockall::automock.

Personally, I'm OK either way. I would only slightly prefer mocking just to be more consistent with the rest of the codebase.

MartinquaXD replied:
I refactored the code to use traits and automock. Although it was possible to test the code with the closures, it made the signature of the optimization function very complex. Also having an automocked trait return different values across multiple invocations is way nicer than make a closure do that.

nlordell commented on crates/solver/src/settlement_post_processing/optimize_unwrapping.rs:
```diff
@@ -0,0 +1,206 @@
+    let mut optimized_settlement = settlement.clone();
+    optimized_settlement.encoder.drop_unwrap(weth.address());
+
+    if settlement_would_succeed(optimized_settlement.clone()).await {
+        tracing::debug!("use internal buffer to unwraps");
+        *settlement = optimized_settlement;
+        return Ok(());
+    }
```
At first, I thought that we could just use the condition `if required_eth_payout <= weth_balance { ... }` here (in order to avoid a node round-trip), but it turns out, you can't because it doesn't capture solvers using `ETH` buffers for some other purpose which could leave not enough `ETH` for the payout at the end. Maybe add a comment here as to why we need a simulation instead of just checking amounts?

MartinquaXD replied:
Done.

nlordell reviewed (commented):
Code looks good! Just not a fan of the `debug_assert`.

Commit:
Drop Itertools::chunk from fn simulating settlements

Commit:
Replace simulation closure with &impl SettlementSimulating

Commit:
Add comments explaining why simulations are required

nlordell commented on crates/solver/src/settlement_post_processing/optimize_unwrapping.rs:
```diff
@@ -0,0 +1,201 @@
+use super::SettlementSimulating;
+use crate::settlement::Settlement;
+use crate::solver::http_solver::buffers::BufferRetrieving;
```
Nice!

nlordell reviewed (approved):
LGTM

Might be cool to do a manual test on Rinkeby and actually submit a settlement to see it use buffers (might require sending some ETH to the settlement contract though).

Commit:
Return unoptimized settlemnt if balance can't be fetched

Commit:
Fix typo in debug log

Commit:
Merge branch 'main' into native-token-unwraps
